js/plugins/googleai/src/gemini.ts

```diff
@@ -279,10 +312,26 @@ export function googleAIModel(
         },
         options
       );
-      const messages = request.messages.map(toGeminiMessage);
+
+      const messages = request.messages;
       if (messages.length === 0) throw new Error('No messages provided.');
+
+      // Gemini does not support messages with role system and instead
+      // expects system instructions to be provided as a separate input.
+      // By convention, system message are expected to be in the "first"
+      // position, so we look for it there. System messages anywhere else
+      // are considered "exceptional".
+      let systemInstruction;
+      if (messages[0].role === 'system') {
+        systemInstruction = toGeminiSystemInstruction(messages[0]);
+      }
```

Comment:
I'd recommend just detecting and splicing off the first system message. Something like:

```ts
const systemMessage = messages.find(m => m.role === 'system');
if (systemMessage) messages.splice(messages.indexOf(systemMessage), 1);
const systemInstruction = toGeminiSystemInstruction(systemMessage);
```

Reply:
How about above you do `const messages = [...request.messages]` to copy the array, then it won't be modified in place.